### PR TITLE
make HTML attribute names case insensitive

### PR DESCRIPTION
**Pull request by Jym77**

### CODE REVIEW

File: packages/alfa-dom/src/node/attribute.ts [hunk 1]
```diff
@@ -1,8 +1,10 @@
-import { None, Option } from "@siteimprove/alfa-option";
+import {None, Option} from "@siteimprove/alfa-option";
+import {Predicate} from "@siteimprove/alfa-predicate";
 
 import { Namespace } from "../namespace";
 import { Node } from "../node";
 import { Element } from "./element";
+import equals = Predicate.equals;
```
Comment (kasperisager) [inline packages/alfa-dom/src/node/attribute.ts]:
```suggestion

const { equals } = Predicate;
```

File: packages/alfa-dom/src/node/attribute.ts [hunk 2]
```diff
@@ -46,7 +48,15 @@ export class Attribute extends Node {
   }
 
   public get name(): string {
-    return this._name;
+    // HTML attributes are case insensitive. Other, e.g. SVG, are case sensitive.
+    return this._namespace.some(equals(Namespace.HTML)) ? this._name.toLowerCase() : this._name;
+  }
+
+  public matchName(name: string): boolean {
```
Comment (kasperisager) [inline packages/alfa-dom/src/node/attribute.ts]:
```suggestion
  public hasName(name: string): boolean {
```

This should also be placed after the getters. We should enforce the ordering using https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/member-ordering.md once ESLint is up and running (#171).

**Comment by kasperisager** [comment]
On a more important note, HTML attributes don't have namespaces so the check should actually be `this._namespace.isNone()`. A safer option could also be to look at the namespace of the owning element.

File: packages/alfa-dom/src/node/attribute.ts [hunk 3]
```diff
@@ -1,8 +1,11 @@
-import { None, Option } from "@siteimprove/alfa-option";
+import {None, Option} from "@siteimprove/alfa-option";
+import {Predicate} from "@siteimprove/alfa-predicate";
 
-import { Namespace } from "../namespace";
-import { Node } from "../node";
-import { Element } from "./element";
+import {Namespace} from "../namespace";
+import {Node} from "../node";
+import {Element} from "./element";
```
Comment (kasperisager) [inline packages/alfa-dom/src/node/attribute.ts]: Something odd happened to the formatting here 🙈

**Comment by Jym77** [comment]
Yeah, it's anyway better to push the computation to the constructor rather than the getter…

Problem is that the constructor can't really call `this.isHTML` (especially if its using `this.owner`). Or at least this sounds like a recipe for disaster… Since `hasName` is still needed (to lowercase the match), this results in some smelly duplication of the test…

File: packages/alfa-dom/src/node/attribute.ts [hunk 4]
```diff
@@ -57,6 +61,13 @@ export class Attribute extends Node {
     return this._owner;
   }
 
+  public hasName(name: string): boolean {
+    // HTML attributes are case insensitive. Other, e.g. SVG, are case sensitive.
+    return this.owner.some(element => element.namespace.some(equals(Namespace.HTML))) ?
+      this._name.toLowerCase() === name.toLowerCase() :
+      this._name === name;
```
Comment (kasperisager) [inline packages/alfa-dom/src/node/attribute.ts]:
```suggestion
    // HTML attributes are case insensitive. Other, e.g. SVG, are case sensitive.
    return this._name ===
      this._owner.some(element => element.namespace.some(equals(Namespace.HTML)))
      ? name.toLowerCase()
      : name;
```

File: packages/alfa-dom/src/node/attribute.ts [hunk 5]
```diff
@@ -32,7 +35,8 @@ export class Attribute extends Node {
 
     this._namespace = namespace;
     this._prefix = prefix;
-    this._name = name;
+    // HTML attributes are case insensitive. Other, e.g. SVG, are case sensitive.
```
Comment (kasperisager) [inline packages/alfa-dom/src/node/attribute.ts]:
```suggestion
    // HTML attributes are case insensitive while attributes in other namespaces aren't.
```

File: packages/alfa-dom/src/node/attribute.ts [hunk 6]
```diff
@@ -57,6 +61,13 @@ export class Attribute extends Node {
     return this._owner;
   }
 
+  public hasName(name: string): boolean {
+    // HTML attributes are case insensitive. Other, e.g. SVG, are case sensitive.
```
Comment (kasperisager) [inline packages/alfa-dom/src/node/attribute.ts]:
```suggestion
    // HTML attributes are case insensitive while attributes in other namespaces aren't.
```

**Comment by kasperisager** [comment]
@Jym77 I'm not too worried about this bit of duplication and do much prefer repeating the same stanza twice over introducing a private method 👍 It's a single line of code anyway 😄

**Pull request merged**